### DISCUSSION TITLE
透明代理下 Xray 内置 DNS Over HTTPS 配置疑问

### DISCUSSION THREAD

**Discussion by ghost** [XTLS/Xray-core#2422] [Q&A] [answered]
看 [白话文指南](https://guide.v2fly.org/basics/dns.html#dns-over-https) 里有这么一段描述：
`需要注意，只有在客户端、服务端都使用 DOH 协议（客户端使用 https 模式，服务端使用 https+local 模式）时候，VPS 出口上才不会出现传统的 UDP DNS 请求。`

那么在透明代理下，例如客户端使用 passwall+SmartDNS（服务端已有 https+local 及相关流程）：
SmartDNS 里的国外 DNS 服务器（使用 passwall socks 代理查询）是否也需要配置为 DOH，才能让 VPS 出口也为 DOH 查询？
还是说普通 UDP 查询（通过 Xray socks 代理）即可满足防污染+VPS端的隐私需求了（还能有速度提升）？

**Comment by chika0801** [comment]
在透明代理环境，以你用的 passwall+SmartDNS 为例，你浏览器发起一个域名的DNS请求，最后这个DNS请求进入了 SmartDNS ，无论你在 SmartDNS 这软件是用国内还是国外DNS查询，假设得到一个IP A。浏览器对这IP A发起请求，这请求进入 Xray core ，由于 passwall 默认设置选项是启用了Xray配置中的域名嗅探参数，Xray 客户端发到你服务端的请求变为了域名。你服务端设它是在国外的，VPS上最终是用UDP还是用DOH查，服务端查完DNS查询返回的IP B，你服务端是访问这IP B。这IP B有可能和你在 SmartDNS 得到的IP A不一样。它通常是离你VPS近的IP。

**Comment by chika0801** [comment]
你 SmartDNS 假设查询这域名用的是国内阿里的DNS，你VPS上查询这域名用的是 谷歌 的DNS，此时当你去那些所谓检查DNS泄露的卖推销VPN的网站，得到的结果是说你DNS泄露。如果你不懂，即使你懂，我也不推荐你关注这DNS泄露的事情。

**Reply by ghost** [reply]
感谢讲解，但我可能没表达清楚，我是想解决VPS端向外部进行DNS查询的隐私，也就是不想让VPS供应商通过我的DNS查询记录，看到我访问了哪些网站。

起因是我偶然发现，我用的某小众VPS（名字就不提了）系统默认DNS并非公共DNS（如8.8.8.8），而是100.64.开头的一个内网IP。当然不是指责他们有监控DNS查询嫌疑，应该只是为了配置某些网站的DNS解锁服务。但总归不太放心，所以就想通过Xray的DNS模块实现：
1. 大部分网站，使用公共DNS进行DOH查询；
2. 少量特殊网站，使用VPS本地DNS查询，以实现解锁服务。

**Reply by ghost** [reply]
服务端应该没什么疑问，配置如下：
```
"dns": {
	"hosts": {
		"dns.google": ["8.8.8.8", "8.8.4.4"]
	},
	"servers": [
		"https+local://1.1.1.1/dns-query",
		"https+local://dns.google/dns-query",
		{
			"address": "localhost",
			"domains": ["geosite:openai"]
		}
	],
	"queryStrategy": "UseIPv4"
}
```
同时开启sniffing和freedom出站UseIPv4。

但客户端就不太清楚该怎么配合（或者是否需要配合）才能解决我说的VPS出口隐私问题。尤其是看到OP里白话文那句：客户端也必须使用DOH协议，VPS出口才不会…才有了疑问。

**Reply by ghost** [reply]
趁着大佬在，多请教一下：
1. 白话文里那句是真的么？我理解DNS请求通过代理发到服务端，具体再往下怎么查询不应该只看服务端的Xray DNS配置么。比如我在客户端访问google.com，DNS为8.8.8.8，于是DNS查询请求被Xray代理，Xray服务端收到请求后，难道不会根据Xray DNS模块配置，将向8.8.8.8的udp明文查询请求，转为向1.1.1.1/dns-query的doh查询？
2. 如果真如白话文所说，那么在透明代理下，通过SmartDNS走passwall socks代理的DNS查询，是否也必须为doh，还是udp 53就够了（毕竟速度快）？
3. sniffing在client或server上，是否只要一端开启就够了？我的理解嗅探完，IP已经变域名了，嗅探两次并没必要？比如我在passwall开了sniffing，那么VPS端其实开不开无所谓（也许关一个还能有性能提升）？或者在server开了sniffing，那么passwall即使开了routeonly也不影响。
4. sniffing是基于什么原理的？从【域名转IP】这种都需要专门的DNS服务，从【IP转域名】不是更麻烦么。比如passwall开了routeonly后，发给VPS的IP代理请求中，难道还是包含了原域名信息，只是被服务端Xray提取出来了？

**Comment by chika0801** [comment]
你服务端这样配是实现你的目的了的

> 大部分网站，使用公共DNS进行DOH查询；
> 少量特殊网站，使用VPS本地DNS查询，以实现解锁服务。

```
"dns": {
	"hosts": {
		"dns.google": ["8.8.8.8", "8.8.4.4"]
	},
	"servers": [
		"https+local://1.1.1.1/dns-query",
		"https+local://dns.google/dns-query",
		{
			"address": "localhost",
			"domains": ["geosite:openai"]
		}
	],
	"queryStrategy": "UseIPv4"
}
```

> 但客户端就不太清楚该怎么配合（或者是否需要配合）才能解决我说的VPS出口隐私问题。尤其是看到OP里白话文那句：客户端也必须使用DOH协议，VPS出口才不会…才有了疑问。

客户端不用配置，工作流程参考 https://github.com/XTLS/Xray-core/discussions/2422#discussioncomment-6680262

**Comment by chika0801** [comment, marked as answer]
> 白话文里那句是真的么？我理解DNS请求通过代理发到服务端，具体再往下怎么查询不应该只看服务端的Xray DNS配置么。比如我在客户端访问google.com，DNS为8.8.8.8，于是DNS查询请求被Xray代理，Xray服务端收到请求后，难道不会根据Xray DNS模块配置，将向8.8.8.8的udp明文查询请求，转为向1.1.1.1/dns-query的doh查询？

假设你客户端对 8.8.8.8 UDP 53发起一个请求，这请求被代理，送到服务端，显然这请求里没域名，域名嗅探不出域名，服务端直接出站访问 8.8.8.8 UDP 53。

这方便问题，你先自学一下 https://tachyondevel.medium.com/%E6%BC%AB%E8%B0%88%E5%90%84%E7%A7%8D%E9%BB%91%E7%A7%91%E6%8A%80%E5%BC%8F-dns-%E6%8A%80%E6%9C%AF%E5%9C%A8%E4%BB%A3%E7%90%86%E7%8E%AF%E5%A2%83%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8-62c50e58cbd0 虽然是老文章，原理基本上没有错，如果你想研究透 HTTPsocks代理环境，透明代理环境，客户端用什么配置，服务端用什么配置 xray/v2ray v4版本 配置中处理流程，看一遍我也没看懂，我慢慢学了好久终于学懂了。

https://guide.v2fly.org/basics/dns.html
https://www.v2fly.org/config/dns.html#dns-%E5%A4%84%E7%90%86%E6%B5%81%E7%A8%8B

这2个链接虽然讲v2ray配置，和Xray现在的原理是一样的，基本没什么大错误。

**Comment by chika0801** [comment]
> 如果真如白话文所说，那么在透明代理下，通过SmartDNS走passwall socks代理的DNS查询，是否也必须为doh，还是udp 53就够了（毕竟速度快）？

参考前面提到的工作原理 `由于 passwall 默认设置选项是启用了Xray配置中的域名嗅探参数，Xray 客户端发到你服务端的请求变为了域名。`

大概解释一下，你用透明代理环境，本地就要处理第1次的DNS请求。假设这不讨论Fakedns。本地DNS后得到的IP，你要选择在客户端发到服务端最终是发这DNS后得到的IP还是嗅探出发域名。如果是发IP，你在VPS端的配置中还要不要开嗅探还原成域名。

所以你要了解懂DNS的流程，又要要求你对 xray 配置熟悉，懂处理流程，综合的一个能力。

**Comment by chika0801** [comment]
> sniffing在client或server上，是否只要一端开启就够了？我的理解嗅探完，IP已经变域名了，嗅探两次并没必要？比如我在passwall开了sniffing，那么VPS端其实开不开无所谓（也许关一个还能有性能提升）？或者在server开了sniffing，那么passwall即使开了routeonly也不影响。

为什么 passwall 作者 包括 ssrp 作者给的生成的xray配置中都默认用了域名嗅探（没开routy only)因为要照顾大部分小白，不可能给每个人解释，只能找一个通用的配置，作者都是义务劳动，不想给小白多解释的道理。当你想研究透，大部分靠自觉，网上问也许运气好遇到愿意给你多说一些的人。或你身边有懂网络知识高手当面别人给你讲。

所以你问 passwall ssrp 为什么作者默认都用了嗅探，有作者的想法。至于服务端我建议你如果不太懂，把入站中都加上嗅探参数（注意没提routy only表示 都不开）都加上没有坏处。

**Comment by chika0801** [comment]
> sniffing是基于什么原理的？

我不懂原理，我只知道用处。

> 比如passwall开了routeonly后，发给VPS的IP代理请求中，难道还是包含了原域名信息，只是被服务端Xray提取出来了？

是的，服务端如果收到的IP，但是开了嗅探，还是能还原成域名。

**Comment by ghost** [comment]
> 假设你客户端对 8.8.8.8 UDP 53发起一个请求，这请求被代理，送到服务端，显然这请求里没域名，域名嗅探不出域名，服务端直接出站访问 8.8.8.8 UDP 53。

明白了…我还以为Xray DNS会“劫持”这个查询请求，没想到直接出站了，怪不得白话文会提到双端DOH。
> https://guide.v2fly.org/basics/dns.html

这么一来，这里讲的 [对外开放 v2ray 的 DNS 服务](https://guide.v2fly.org/basics/dns.html#%E5%AF%B9%E5%A4%96%E5%BC%80%E6%94%BE-v2ray-%E7%9A%84-dns-%E6%9C%8D%E5%8A%A1) 功能也开始理解了。

再次感谢耐心解答，也感谢大佬们花时间写的文档 :bow:

**Reply by chika0801** [reply]
你电脑上把dns设置成127.0.0.1，你xray配置监听本地53端口，dns流量就进来了。此时你电脑相对你vps，算成客户端。vps的dns也可以填127.0.0.1，这是一个最简单的举例

**Comment by chika0801** [comment]
你不要关讨论区帖子，关了更没人看得到了。你选个答案就行
